Add Apache Kafka Broker (#2740)
knative/docs: docs/eventing/broker/kafka-broker.md

---
title: "Apache Kafka Broker"
weight: 30
type: "docs"
---

This page shows how to install and configure the 
[Kafka Broker](https://github.com/knative-sandbox/eventing-kafka-broker).

This guide assumes a 
[Knative Eventing installation](./../../install/any-kubernetes-cluster.md#installing-the-eventing-component).

## Installation

<!-- TODO Where does the config for populating the following live? -->
```bash
kubectl apply --filename {{< artifact repo="eventing-kafka-broker" file="eventing-kafka-broker.yaml" >}}
```
## Create a Kafka Broker

A Kafka Broker object looks like this:

```yaml
apiVersion: eventing.knative.dev/v1
kind: Broker
metadata:
  annotations:
    # case-sensitive
    eventing.knative.dev/broker.class: kafka 
  name: default
  namespace: default
spec:
  # Configuration specific to this broker.
  config:
    apiVersion: v1
    kind: ConfigMap
    name: kafka-broker-config
    namespace: knative-eventing
  # Optional dead letter sink, you can specify either:

>   #  - deadLetterSink.ref, which is a reference to a Callable

Reviewer: I think you should not spend time here explaining how to configure the deadLetterSink, that's something that should be covered in the broker docs itself

Author: No, strong opinion, I thought it was nice to have a full Broker definition for this particular implementation.

```

The above `ConfigMap` is installed in the cluster, and you can edit
the configurations or create a new one with the same shape 
depending on your needs.

## Set as default broker implementation

To set the Kafka Broker as your default implementation you can edit
the `config-br-defaults` `ConfigMap` in `knative-eventing` namespace.
In this way, you can avoid specifying 
`metadata.annotations.eventing.knative.dev/broker.class` or 
`spec.config` for each Kafka Broker created in the cluster
or in a specific namespace.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-br-defaults
  namespace: knative-eventing
data:
  default-br-config: |
    clusterDefault:
      brokerClass: kafka 
      apiVersion: v1
      kind: ConfigMap
      name: kafka-broker-config
      namespace: knative-eventing
    namespaceDefaults:
      namespace1:
        brokerClass: kafka 
        apiVersion: v1
        kind: ConfigMap
        name: kafka-broker-config
        namespace: knative-eventing
      namespace2:
        brokerClass: kafka 
        apiVersion: v1
        kind: ConfigMap
        name: kafka-broker-config
        namespace: knative-eventing
```

> 

Reviewer: Should you spend a bunch of more words about the other CMs (logging, kafka properties, etc?)

Author: Yeah, let me do that

Author: Done: https://github.com/knative/docs/pull/2740/commits/cb6dc488fc3f7d7e239aec72954adf040fda4676

---
title: "Apache Kafka Broker"
weight: 30
type: "docs"
---

This page shows how to install and configure the 
[Kafka Broker](https://github.com/knative-sandbox/eventing-kafka-broker).

This guide assumes a 
[Knative Eventing installation](./../../install/any-kubernetes-cluster.md#installing-the-eventing-component).

## Installation

<!-- TODO Where does the config for populating the following live? -->
```bash

> kubectl apply --filename {{< artifact repo="eventing-kafka-broker" file="eventing-kafka-broker.yaml" >}}

Author: Do we need a verification step below?

---
title: "Apache Kafka Broker"
weight: 30
type: "docs"
---

This page shows how to install and configure the 

> [Kafka Broker](https://github.com/knative-sandbox/eventing-kafka-broker).

Reviewer 2: What's the user story for this, i.e. why would someone use this over the default Knative broker?
This would probably be a good place for some introductory content, not so much about what the Kafka broker is (we could include the link from earlier here again maybe), but _why_ as a user would you want to install or use this?

Author: I added this https://github.com/knative/docs/pull/2740/files#diff-20631ccdd823c2921c4d6bda9b486bfaR7-R17, please, let me know if it helps.

> 

Reviewer 2: ```suggestion
## Prerequisites
```

This guide assumes a 

> [Knative Eventing installation](./../../install/any-kubernetes-cluster.md#installing-the-eventing-component), only
> `eventing-core` is required.

Reviewer 2: ```suggestion
[Knative Eventing installation](./../../install/any-kubernetes-cluster.md#installing-the-eventing-component).
```

Reviewer 2: I'm not sure if it's necessary to say just `eventing-core` is required, we assume they'll install Knative Eventing as is?

Author: I wanted to point out that a user doesn't need to install everything listed there but only the core, ie a user can stop the installation to the second bullet of this section: https://knative.dev/development/install/any-kubernetes-cluster/#installing-the-eventing-component

## Installation

> ```bash
> kubectl apply --filename {{< artifact org="knative-sandbox" repo="eventing-kafka-broker" file="eventing-kafka-broker.yaml" >}}
> ```

Reviewer 2: ```suggestion
Install the Kafka broker by entering the following command:

```bash
kubectl apply --filename {{< artifact org="knative-sandbox" repo="eventing-kafka-broker" file="eventing-kafka-broker.yaml" >}}
```
```

> Verify that `kafka-broker-controller`, `kafka-broker-receiver` and `kafka-broker-dispatcher` are running:
> 
> ```bash
> kubectl get deployments.apps -n knative-eventing 
> 
> NAME                           READY   UP-TO-DATE   AVAILABLE   AGE
> eventing-controller            1/1     1            1           10s
> eventing-webhook               1/1     1            1           9s
> kafka-broker-controller        1/1     1            1           3s
> kafka-broker-dispatcher        1/1     1            1           4s
> kafka-broker-receiver          1/1     1            1           5s
> ```

Reviewer 2: ```suggestion
2. Verify that `kafka-broker-controller`, `kafka-broker-receiver` and `kafka-broker-dispatcher` are running, by entering the following command:

```bash
kubectl get deployments.apps -n knative-eventing 
```

Example output:

```
NAME                           READY   UP-TO-DATE   AVAILABLE   AGE
eventing-controller            1/1     1            1           10s
eventing-webhook               1/1     1            1           9s
kafka-broker-controller        1/1     1            1           3s
kafka-broker-dispatcher        1/1     1            1           4s
kafka-broker-receiver          1/1     1            1           5s
```
```

Reviewer 2: idk why when I try to suggest it it's rendering weird, but above is how I'd recommend laying this out (you'd need to indent the output properly too, but as a separate code block to the command).

Author: Done: https://github.com/knative/docs/pull/2740/commits/c8969cb556b1dad975a66a0688e7e85d5d985e04

## Create a Kafka Broker

A Kafka Broker object looks like this:

```yaml
apiVersion: eventing.knative.dev/v1
kind: Broker
metadata:
  annotations:
    # case-sensitive
    eventing.knative.dev/broker.class: Kafka
  name: default
  namespace: default
spec:
  # Configuration specific to this broker.
  config:
    apiVersion: v1
    kind: ConfigMap
    name: kafka-broker-config
    namespace: knative-eventing
  # Optional dead letter sink, you can specify either:
  #  - deadLetterSink.ref, which is a reference to a Callable
  #  - deadLetterSink.uri, which is an absolute URI to a Callable (It can potentially be out of the Kubernetes cluster)
  delivery: 
    deadLetterSink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: dlq-service

> ```

Reviewer 2: Move this to the beginning of the doc instead and use it to explain what's required and why in a kafka broker

Author: Before the installation?

`spec.config` should reference any `ConfigMap` that looks like the following:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config
  namespace: knative-eventing
data:
  # Number of topic partitions
  default.topic.partitions: "10"
  # Replication factor of topic messages.
  default.topic.replication.factor: "1" 
  # A comma separated list of bootstrap servers. (It can be in or out the k8s cluster)
  bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092" 
```

The above `ConfigMap` is installed in the cluster, and you can edit
the configurations or create a new one with the same shape 

> depending on your needs.

Reviewer 2: Let's be more prescriptive than this. Is it specifically yhe `kafka-broker-config` ConfigMap that needs to be referenced?
If we're going to talk about a previously unmentioned ConfigMap in the docs, it should probably be explained properly in its own section, as it is for the default broker: https://knative.dev/development/eventing/broker/config-br-defaults/

Author: Are you suggesting to create another page for this config map?

## Set as default broker implementation

> To set the Kafka Broker as your default implementation you can edit
> the `config-br-defaults` `ConfigMap` in `knative-eventing` namespace.
> In this way, you can avoid specifying 
> `metadata.annotations.eventing.knative.dev/broker.class` or 
> `spec.config` for each Kafka Broker created in the cluster
> or in a specific namespace.

Reviewer 2: ```suggestion
To set the Kafka broker as the default implementation for all brokers in the Knative deployment, you can apply global settings by modifying the `config-br-defaults` ConfigMap in the `knative-eventing` namespace.

This allows you to avoid configuring individual or per-namespace settings for each broker, such as `metadata.annotations.eventing.knative.dev/broker.class` or 
`spec.config`.
```

> 

Reviewer 2: Add a line explaining what YAMLs like this are, e.g. "The following YAML is an example of a `config-br-defaults` ConfigMap using Kafka broker as the default implementation."

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-br-defaults
  namespace: knative-eventing
data:
  default-br-config: |
    clusterDefault:
      brokerClass: kafka 
      apiVersion: v1
      kind: ConfigMap
      name: kafka-broker-config
      namespace: knative-eventing
    namespaceDefaults:
      namespace1:
        brokerClass: kafka 
        apiVersion: v1
        kind: ConfigMap
        name: kafka-broker-config
        namespace: knative-eventing
      namespace2:
        brokerClass: kafka 
        apiVersion: v1
        kind: ConfigMap
        name: kafka-broker-config
        namespace: knative-eventing
```

> ## Change Kafka Producer and Consumer configuration
> 
> We expose all Kafka producer and consumer configurations, so that they can be 
> changed accordingly to specific workloads that it needs to run.
> 
> Such configurations are in the `ConfigMap` `config-kafka-broker-data-plane`
> in the `knative-eventing` namespace. All configurations in that `ConfigMap`
> are documented on the [Apache Kafka website](https://kafka.apache.org/documentation/),
> in particular, [Producer configurations](https://kafka.apache.org/documentation/#producerconfigs)
> and [Consumer configurations](https://kafka.apache.org/documentation/#consumerconfigs).

Reviewer 2: ```suggestion
## Kafka Producer and Consumer configurations

Knative exposes all available Kafka producer and consumer configurations that can be modified to suit your workloads.

You can change these configurations by modifying the `config-kafka-broker-data-plane` ConfigMap in the `knative-eventing` namespace.

Documentation for the settings available in this ConfigMap is available on the [Apache Kafka website](https://kafka.apache.org/documentation/),
in particular, [Producer configurations](https://kafka.apache.org/documentation/#producerconfigs)
and [Consumer configurations](https://kafka.apache.org/documentation/#consumerconfigs).
```

> ## Change logging level to data plane components

Reviewer 2: ```suggestion
## Enable debug data logging for data plane components
```

> The logging config for the data plane components looks like this:

Reviewer 2: ```suggestion
The default logging configuration for Kafka data plane components looks like this:
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config-logging
  namespace: knative-eventing
data:
  config.xml: |
    <configuration>
      <appender name="jsonConsoleAppender" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LogstashEncoder"/>
      </appender>
      <root level="INFO">
        <appender-ref ref="jsonConsoleAppender"/>
      </root>
    </configuration>
```

> To change the log level to `DEBUG`, you need to apply the following `ConfigMap`:

Reviewer 2: ```suggestion
To change the logging level to `DEBUG`, you must apply the following `kafka-broker-config-logging` ConfigMap.
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config-logging
  namespace: knative-eventing
data:
  config.xml: |
    <configuration>
      <appender name="jsonConsoleAppender" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LogstashEncoder"/>
      </appender>
      <root level="DEBUG">
        <appender-ref ref="jsonConsoleAppender"/>
      </root>
    </configuration>

> ```

Reviewer 2: Do they need to create this from scratch or update an existing file?

Author: Update and existing `ConfigMap`, I clarified this point more.

We don't support hot-reload for logging configuration, because it can be
unnecessary, but if there is much interest we can add support for it, so,
in this case we need to restart data plane components by running the 
following commands:

```bash
kubectl rollout restart deployment -n knative-eventing kafka-broker-receiver
kubectl rollout restart deployment -n knative-eventing kafka-broker-dispatcher

> ```

Reviewer 2: Remove anything not supported from docs, even if there's a hack for it or it might be supported in the future.

### Additional information

> - For reporting bugs or feature requests open an issue to the repository: 
> [eventing-kafka-broker](https://github.com/knative-sandbox/eventing-kafka-broker) 

Reviewer 2: ```suggestion
- To report bugs or add feature requests, open an issue in the [eventing-kafka-broker repository](https://github.com/knative-sandbox/eventing-kafka-broker).
```

  #  - deadLetterSink.uri, which is an absolute URI to a Callable (It can potentially be out of the Kubernetes cluster)
  delivery:
    deadLetterSink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: dlq-service
```

`spec.config` should reference any `ConfigMap` that looks like the following:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config
  namespace: knative-eventing
data:
  # Number of topic partitions
  default.topic.partitions: "10"
  # Replication factor of topic messages.
  default.topic.replication.factor: "1"
  # A comma separated list of bootstrap servers. (It can be in or out the k8s cluster)
  bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092"
```

The above `ConfigMap` is installed in the cluster, and you can edit
the configurations or create a new one with the same shape
depending on your needs.

## Set as default broker implementation

To set the Kafka broker as the default implementation for all brokers in the Knative deployment, you can apply global settings by modifying the `config-br-defaults` ConfigMap in the `knative-eventing` namespace.

> This allows you to avoid configuring individual or per-namespace settings for each broker, such as `metadata.annotations.eventing.knative.dev/broker.class` or 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
This allows you to avoid configuring individual or per-namespace settings for each broker, such as `metadata.annotations.eventing.knative.dev/broker.class` or
```

> This allows you to avoid configuring individual or per-namespace settings for each broker, such as `metadata.annotations.eventing.knative.dev/broker.class` or 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
This allows you to avoid configuring individual or per-namespace settings for each broker, such as `metadata.annotations.eventing.knative.dev/broker.class` or
```

`spec.config`.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-br-defaults
  namespace: knative-eventing
data:
  default-br-config: |
    clusterDefault:

>       brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
      brokerClass: kafka
```

>       brokerClass: Kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
      brokerClass: Kafka
```

      apiVersion: v1
      kind: ConfigMap
      name: kafka-broker-config
      namespace: knative-eventing
    namespaceDefaults:
      namespace1:

>         brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: kafka
```

>         brokerClass: Kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: Kafka
```

        apiVersion: v1
        kind: ConfigMap
        name: kafka-broker-config
        namespace: knative-eventing
      namespace2:

>         brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: kafka
```

>         brokerClass: Kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: Kafka
```

---
title: "Apache Kafka Broker"
weight: 30
type: "docs"
---

> This page shows how to install and configure the 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
This page shows how to install and configure the
```

> This page shows how to install and configure the 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
This page shows how to install and configure the
```

    ```bash
    kubectl get deployments.apps -n knative-eventing
    ```

    Example output:

    ```bash
    NAME                           READY   UP-TO-DATE   AVAILABLE   AGE
    eventing-controller            1/1     1            1           10s
    eventing-webhook               1/1     1            1           9s
    kafka-broker-controller        1/1     1            1           3s
    kafka-broker-dispatcher        1/1     1            1           4s
    kafka-broker-receiver          1/1     1            1           5s
    ```

## Create a Kafka Broker

A Kafka Broker object looks like this:

```yaml
apiVersion: eventing.knative.dev/v1
kind: Broker
metadata:
  annotations:
    # case-sensitive
    eventing.knative.dev/broker.class: Kafka
  name: default
  namespace: default
spec:
  # Configuration specific to this broker.
  config:
    apiVersion: v1
    kind: ConfigMap
    name: kafka-broker-config
    namespace: knative-eventing
  # Optional dead letter sink, you can specify either:
  #  - deadLetterSink.ref, which is a reference to a Callable
  #  - deadLetterSink.uri, which is an absolute URI to a Callable (It can potentially be out of the Kubernetes cluster)

>   delivery: 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
  delivery:
```

>   delivery: 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
  delivery:
```

    deadLetterSink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: dlq-service
```

`spec.config` should reference any `ConfigMap` that looks like the following:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config
  namespace: knative-eventing
data:
  # Number of topic partitions
  default.topic.partitions: "10"
  # Replication factor of topic messages.

>   default.topic.replication.factor: "1" 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
  default.topic.replication.factor: "1"
```

>   default.topic.replication.factor: "1" 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
  default.topic.replication.factor: "1"
```

  # A comma separated list of bootstrap servers. (It can be in or out the k8s cluster)

>   bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092" 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
  bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092"
```

>   bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092" 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
  bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092"
```

```

The above `ConfigMap` is installed in the cluster, and you can edit

> the configurations or create a new one with the same shape 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
the configurations or create a new one with the same shape
```

> the configurations or create a new one with the same shape 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
the configurations or create a new one with the same shape
```

depending on your needs.

## Set as default broker implementation

To set the Kafka Broker as your default implementation you can edit
the `config-br-defaults` `ConfigMap` in `knative-eventing` namespace.

> In this way, you can avoid specifying 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
In this way, you can avoid specifying
```

> In this way, you can avoid specifying 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
In this way, you can avoid specifying
```

> `metadata.annotations.eventing.knative.dev/broker.class` or 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
`metadata.annotations.eventing.knative.dev/broker.class` or
```

> `metadata.annotations.eventing.knative.dev/broker.class` or 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
`metadata.annotations.eventing.knative.dev/broker.class` or
```

`spec.config` for each Kafka Broker created in the cluster
or in a specific namespace.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-br-defaults
  namespace: knative-eventing
data:
  default-br-config: |
    clusterDefault:

>       brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
      brokerClass: kafka
```

>       brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
      brokerClass: kafka
```

      apiVersion: v1
      kind: ConfigMap
      name: kafka-broker-config
      namespace: knative-eventing
    namespaceDefaults:
      namespace1:

>         brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: kafka
```

>         brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: kafka
```

        apiVersion: v1
        kind: ConfigMap
        name: kafka-broker-config
        namespace: knative-eventing
      namespace2:

>         brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: kafka
```

>         brokerClass: kafka 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
        brokerClass: kafka
```

The logging config for the data plane components looks like this:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config-logging
  namespace: knative-eventing
data:
  config.xml: |
    <configuration>
      <appender name="jsonConsoleAppender" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LogstashEncoder"/>
      </appender>
      <root level="INFO">
        <appender-ref ref="jsonConsoleAppender"/>
      </root>
    </configuration>
```

To change the log level to `DEBUG`, you need to apply the following `ConfigMap`:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config-logging
  namespace: knative-eventing
data:
  config.xml: |
    <configuration>
      <appender name="jsonConsoleAppender" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LogstashEncoder"/>
      </appender>
      <root level="DEBUG">
        <appender-ref ref="jsonConsoleAppender"/>
      </root>
    </configuration>
```

We don't support hot-reload for logging configuration, because it can be
unnecessary, but if there is much interest we can add support for it, so,

> in this case we need to restart data plane components by running the 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
in this case we need to restart data plane components by running the
```

> in this case we need to restart data plane components by running the 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
in this case we need to restart data plane components by running the
```

following commands:

```bash
kubectl rollout restart deployment -n knative-eventing kafka-broker-receiver
kubectl rollout restart deployment -n knative-eventing kafka-broker-dispatcher
```

### Additional information

> - For reporting bugs or feature requests open an issue to the repository: 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
- For reporting bugs or feature requests open an issue to the repository:
```

> - For reporting bugs or feature requests open an issue to the repository: 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
- For reporting bugs or feature requests open an issue to the repository:
```

> [eventing-kafka-broker](https://github.com/knative-sandbox/eventing-kafka-broker) 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
[eventing-kafka-broker](https://github.com/knative-sandbox/eventing-kafka-broker)
```

> [eventing-kafka-broker](https://github.com/knative-sandbox/eventing-kafka-broker) 

Reviewer 3: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
[eventing-kafka-broker](https://github.com/knative-sandbox/eventing-kafka-broker)
```

kind: Broker
metadata:
  annotations:
    # case-sensitive
    eventing.knative.dev/broker.class: Kafka
  name: default
  namespace: default
spec:
  # Configuration specific to this broker.
  config:
    apiVersion: v1
    kind: ConfigMap
    name: kafka-broker-config
    namespace: knative-eventing
  # Optional dead letter sink, you can specify either:
  #  - deadLetterSink.ref, which is a reference to a Callable
  #  - deadLetterSink.uri, which is an absolute URI to a Callable (It can potentially be out of the Kubernetes cluster)
  delivery:
    deadLetterSink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: dlq-service
```

`spec.config` should reference any `ConfigMap` that looks like the following:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config
  namespace: knative-eventing
data:
  # Number of topic partitions
  default.topic.partitions: "10"
  # Replication factor of topic messages.
  default.topic.replication.factor: "1"
  # A comma separated list of bootstrap servers. (It can be in or out the k8s cluster)
  bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092"
```

> The above `ConfigMap` is installed in the cluster, and you can edit

Reviewer 4: ```suggestion
The above `ConfigMap` is installed in the cluster. You can edit
```

> the configurations or create a new one with the same shape

Reviewer 4: ```suggestion
the configuration or create a new one with the same values
```

values or maybe information instead of shape, I think either reads better.
